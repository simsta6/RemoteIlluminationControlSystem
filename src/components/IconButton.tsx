import React from "react";
import {
    GestureResponderEvent,
    StyleProp, TouchableOpacity,
    ViewStyle
} from "react-native";

interface Props {
    onPress?: ((event: GestureResponderEvent) => void) | undefined;
    buttonStyle?: StyleProp<ViewStyle>;
    Icon?: () => JSX.Element;
}

export const IconButton = (props: Props) => {
    const { onPress, buttonStyle, Icon } = props;

    return (
        <TouchableOpacity 
            onPress={onPress}
            style={buttonStyle}
        >
            {Icon && <Icon/>}
        </TouchableOpacity>
    );
};
