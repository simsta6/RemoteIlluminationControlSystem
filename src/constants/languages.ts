export const LANGUAGES = [
    { 
        code: "en", 
        label: "English", 
    },
    { 
        code: "lt", 
        label: "Lietuvi≈≥", 
    }
];