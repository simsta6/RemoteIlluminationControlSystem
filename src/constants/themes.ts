import { Theme } from "@react-navigation/native";

export const KEY_FOR_STORAGE = "defaultScheme";

interface Colors {
    primary: string;
    background: string;
    card: string;
    text: string;
    disabledText: string;
    icon: string;
    border: string;
    notification: string;
    modal: string;
    button: string;
    disabledButton: string;
}

export const AppDarkColors: Colors = {
    primary: "rgba(133, 170, 164, 1)",
    background: "rgba(48, 48, 48, 1)",
    card: "rgba(33, 33, 33, 1)",
    text: "rgba(255, 228, 234, 1)", 
    disabledText: "rgba(186, 168, 172, 1)", 
    icon: "rgba(255, 228, 234, 1)", 
    border: "rgba(51, 70, 69, 1)",
    notification: "rgba(76, 96, 68, 1)",
    modal: "rgba(66, 66, 66, 1)",
    button: "rgba(33, 33, 33, 1)",
    disabledButton: "rgba(74, 74, 74, 1)",
};

export const AppLightColors: Colors = {
    primary: "rgba(122, 85, 91, 1)",
    background: "rgba(255, 235, 236, 1)",
    card: "rgba(229, 208, 237, 1)",
    text: "rgba(122, 85, 91, 1)", 
    disabledText: "rgba(87, 74, 76, 1)", 
    icon: "rgba(122, 85, 91, 1)", 
    border: "rgba(204, 185, 186, 1)",
    notification: "rgba(179, 159, 187, 1)",
    modal: "rgba(255, 247, 248, 1)",
    button: "rgba(229, 208, 237, 1)",
    disabledButton: "rgba(171, 163, 173, 1)",
};

export const AppDarkTheme: Theme = {
    dark: true,
    colors: AppDarkColors,
};

export const AppLightTheme: Theme = {
    dark: false,
    colors: AppLightColors
};