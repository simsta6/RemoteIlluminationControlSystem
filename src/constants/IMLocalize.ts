import i18n, { LanguageDetectorAsyncModule } from "i18next";
import { initReactI18next } from "react-i18next";
import AsyncStorage from "@react-native-async-storage/async-storage";
import * as RNLocalize from "react-native-localize";
import { LANGUAGES } from "./languages";
import en from "../translations/en.json";
import lt from "../translations/lt.json";


const LANG_CODES = Object.keys(LANGUAGES.map(lang => lang.code));

const LANGUAGE_DETECTOR: LanguageDetectorAsyncModule = {
    type: "languageDetector",
    async: true,
    detect: (callback) => {
        AsyncStorage.getItem("user-language", (err, language) => {
            if (err || !language) {
                if (err) {
                    console.log("Error fetching Languages from asyncstorage ", err);
                } else {
                    console.log("No language is set, choosing English as fallback");
                }
                const findBestAvailableLanguage = RNLocalize.findBestAvailableLanguage(LANG_CODES);
  
                callback(findBestAvailableLanguage?.languageTag || "en");
                return;
            }
            callback(language);
        });
    },
    init: () => { 
        // later
    },
    cacheUserLanguage: (language: string) => {
        AsyncStorage.setItem("user-language", language);
    }
};

const resources = {
    en,
    lt,
} as const;

i18n
    .use(LANGUAGE_DETECTOR)
    .use(initReactI18next)
    .init({
        compatibilityJSON: "v3",
        lng: "en",
        react: {
            useSuspense: false
        },
        interpolation: {
            escapeValue: false
        },
        resources,
    });