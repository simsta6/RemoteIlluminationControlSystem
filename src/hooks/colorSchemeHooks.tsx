import { useTheme } from "@react-navigation/native";
import { AppDarkColors, AppLightColors } from "../constants/themes";

export const useAppColors = () => {
    const { dark } = useTheme();

    return {
        colors: dark ? AppDarkColors : AppLightColors
    };
};
