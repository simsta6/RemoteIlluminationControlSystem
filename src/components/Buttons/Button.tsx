import React from "react";
import { 
    GestureResponderEvent, 
    StyleProp, 
    StyleSheet, 
    Text, 
    TextStyle, 
    TouchableOpacity, 
    ViewStyle
} from "react-native";
import { useAppColors } from "../../hooks/colorSchemeHooks";

interface Props {
    title: string;
    onPress?: ((event: GestureResponderEvent) => void) | undefined;
    disabled?: boolean;
    buttonStyle?: StyleProp<ViewStyle>;
    textStyle?: StyleProp<TextStyle>;
}

export const Button = (props: Props) => {
    const { title, onPress, disabled, buttonStyle, textStyle } = props;
    const { colors } = useAppColors();

    return (
        <TouchableOpacity 
            testID={"ButtonTouchableOpacity"}
            style={[{...styles.button, backgroundColor: disabled ? colors.disabledButton : colors.button }, buttonStyle]} 
            onPress={onPress}
            disabled={disabled}
        >
            <Text 
                testID={"ButtonText"}
                style={[{...styles.buttonText, color: disabled ? colors.disabledText : colors.text}, textStyle]}
            >
                { title }
            </Text>
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    button: {
        alignItems: "center",
        margin: 4,
        padding: 8,  
        borderRadius: 2,
        marginVertical: 5,
    },
    buttonText: {
        textTransform: "uppercase",
        textAlign: "center",
        fontWeight: "500",
    }
});