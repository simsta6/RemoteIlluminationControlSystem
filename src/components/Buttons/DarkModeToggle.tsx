import React from "react";
import { useAppColors } from "../../hooks/colorSchemeHooks";
import { useTheme } from "../../hooks/themesHooks";
import { Switch } from "react-native-switch";
import MoonIcon from "../../assets/icons/MoonIcon";
import SunIcon from "../../assets/icons/SunIcon";
import { StyleSheet, Text, View } from "react-native";
import { useTranslation } from "react-i18next";

export const DarkModeToggle = () => {
    const { t } = useTranslation();
    const { colors } = useAppColors();
    const [scheme, actions] = useTheme();
    const [isEnabled, setIsEnabled] = React.useState(scheme === "dark");

    const toggleSwitch = () => {
        setIsEnabled(previousState => !previousState);
    };

    React.useEffect(() => {
        isEnabled ? actions.toDarkTheme() : actions.toLightTheme();
    }, [isEnabled]);

    return (            
        <>
            <Text style={{...styles.title, color: colors.text}}>{t("DarkModeToggle:darkModeToggle")}</Text>
            <View style={styles.container} testID={"Container"}>
                <Switch
                    renderInsideCircle={() => scheme === "dark" ? 
                        <MoonIcon height={25} width={25} color={colors.icon} /> : 
                        <SunIcon height={25} width={25} color={colors.icon} />
                    }
                    testID={"Switch"}
                    backgroundActive={colors.card}
                    backgroundInactive={colors.card}
                    circleActiveColor={colors.background}
                    circleInActiveColor={colors.background}
                    activeText={""}
                    inActiveText={""}
                    onValueChange={toggleSwitch}
                    value={isEnabled}
                    circleSize={30}
                    circleBorderActiveColor={colors.border}
                    circleBorderInactiveColor={colors.border}
                />
            </View>
        </>
    );
};

const styles = StyleSheet.create({
    container: {
        paddingTop: 10,
    },
    title: {
        fontSize: 20,
        fontWeight: "500"
    },
});