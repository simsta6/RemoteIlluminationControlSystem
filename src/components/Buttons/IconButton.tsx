import React from "react";
import {
    GestureResponderEvent,
    StyleProp, StyleSheet, Text, TextStyle, TouchableOpacity,
    View,
    ViewStyle
} from "react-native";
import { Colors } from "../../constants/themes";
import { useAppColors } from "../../hooks/colorSchemeHooks";

const getText = (
    title: string, 
    disabled: boolean | undefined, 
    colors: Colors, 
    textStyle: StyleProp<TextStyle> | undefined
) => {
    return (
        <Text 
            style={[{...styles.buttonText, color: disabled ? colors.disabledText : colors.text}, textStyle]}
        >
            { title }
        </Text>
    );
}; 

interface Props {
    onPress?: ((event: GestureResponderEvent) => void) | undefined;
    buttonStyle?: StyleProp<ViewStyle>;
    Icon: () => JSX.Element;
    title?: string;
    textStyle?: StyleProp<TextStyle>;
    disabled?: boolean;
}

export const IconButton = (props: Props) => {
    const { onPress, buttonStyle, Icon, title, disabled, textStyle } = props;
    const { colors } = useAppColors();

    return (
        <TouchableOpacity 
            onPress={onPress}
            style={buttonStyle}
        >
            <View style={styles.row}>
                {title && getText(title, disabled, colors, textStyle)}
                <Icon/>
            </View>
        </TouchableOpacity>
    );
};

const styles = StyleSheet.create({
    row: {
        flexDirection: "row",
        alignItems: "center"
    },
    buttonText: {
        textAlign: "center",
        fontWeight: "500",
    }
});