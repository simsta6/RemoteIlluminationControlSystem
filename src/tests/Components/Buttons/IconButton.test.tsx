import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { Text } from "react-native";
import { IconButton } from "../../../components/Buttons/IconButton";
import { Colors } from "../../../constants/themes";
import * as hooks from "../../../hooks/colorSchemeHooks";

describe("<IconButton />", () => {
    const colors = {
        disabledText: "orange",
        text: "purple",
    };

    beforeAll(() => {
        jest.spyOn(hooks, "useAppColors").mockImplementation(() => ({ colors: colors as Colors }));
    });

    test("Default IconButton should match snapshot", () => {
        const { toJSON } = render(
            <IconButton Icon={() => <Text>TestIcon</Text>} />
        );

        expect(toJSON()).toMatchSnapshot();
    });

    test("Disabled IconButton should be in rendered correctly", () => {
        const { getByTestId, getByText } = render(
            <IconButton Icon={() => <Text>TestIcon</Text>} title={"TestTitle"} disabled/>
        );

        const touchableOpacity = getByTestId("TouchableOpacityIconButton");
        const text = getByTestId("IconButtonText");

        expect(touchableOpacity.props.accessibilityState.disabled).toBe(true);
        expect(text.props.style[0].color).toBe(colors.disabledText);
        expect(getByText("TestTitle")).toBeTruthy;
    });

    test("onPress function should be called once when button is pressed", () => {
        const mockFn = jest.fn();

        const { getByTestId } = render(
            <IconButton Icon={() => <Text>TestIcon</Text>} onPress={mockFn} />
        );

        fireEvent.press(getByTestId("TouchableOpacityIconButton"));

        expect(mockFn).toBeCalledTimes(1);
    });

    test("onPress function should not be called when Button is disabled", () => {
        const mockFn = jest.fn();

        const { getByTestId } = render(
            <IconButton Icon={() => <Text>TestIcon</Text>} disabled onPress={mockFn} />
        );

        fireEvent.press(getByTestId("TouchableOpacityIconButton"));

        expect(mockFn).toBeCalledTimes(0);
    });

    test("Button text should be the same as given", () => {
        const { getByText } = render(
            <IconButton Icon={() => <Text>TestIcon</Text>} title="IconButtonTestText" />
        );

        expect(getByText("IconButtonTestText").props.children).toBe("IconButtonTestText");
    });

    test("Given styles should be present", () => {
        const { getByTestId } = render(
            <IconButton title={"TestTitle"} buttonStyle={{ height: 100, width: 100 }} textStyle={{ height: 100, width: 100 }} Icon={() => <Text>TestIcon</Text>} />
        );

        const text = getByTestId("IconButtonText");
        const touchableOpacity = getByTestId("TouchableOpacityIconButton");

        expect(text.props.style[1].height).toBe(100);
        expect(text.props.style[1].width).toBe(100);
        expect(touchableOpacity.props.style.height).toBe(100);
        expect(touchableOpacity.props.style.width).toBe(100);
    });
});