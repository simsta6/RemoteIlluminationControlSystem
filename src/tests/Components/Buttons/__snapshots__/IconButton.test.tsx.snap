// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> Default IconButton should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="TouchableOpacityIconButton"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text>
      TestIcon
    </Text>
  </View>
</View>
`;
